### ISSUE TITLE
short dynamics hairpin angle limit

### ISSUE THREAD

**Issue by craigsapp** [rism-ch/verovio#439]
Short hairpin dynamics in verovio currently have the same height as long hairpin dynamics.  It would be better if the height of short hairpins were limited to a maximum angle:

<img width="329" alt="screen shot 2017-01-17 at 20 21 06" src="https://cloud.githubusercontent.com/assets/3487289/22050801/d0be0c7a-dcf2-11e6-99c2-e341317fbccb.png">

In this case the short hairpin on the eighth notes opens too wide given the length of the hairpin.

MEI data:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="http://music-encoding.org/schema/3.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="http://music-encoding.org/schema/3.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="3.0.0">
    <meiHead>
        <fileDesc>
            <titleStmt>
                <title />
            </titleStmt>
            <pubStmt>
                <date>2017-01-17 20:20:28</date>
            </pubStmt>
        </fileDesc>
        <encodingDesc>
            <projectDesc>
                <p>Transcoded from Humdrum with Verovio version 0.9.14-dev-fe01685</p>
            </projectDesc>
        </encodingDesc>
    </meiHead>
    <music>
        <body>
            <mdiv>
                <score>
                    <scoreDef xml:id="scoredef-000000201540999">
                        <staffGrp xml:id="m-000000150502018">
                            <staffDef xml:id="staffdef-000000183058674" clef.shape="G" clef.line="2" meter.count="4" meter.unit="4" n="1" lines="5" />
                        </staffGrp>
                    </scoreDef>
                    <section xml:id="section-000000015777043">
                        <measure xml:id="measure-L3" n="1">
                            <staff xml:id="staff-L3F1" n="1">
                                <layer xml:id="layer-L3F1" n="1">
                                    <note xml:id="note-L4F1" dur="1" oct="4" pname="c" accid.ges="n" />
                                </layer>
                            </staff>
                            <hairpin xml:id="hairpin-000000110600477" staff="1" tstamp="1.000000" tstamp2="1m+1.0" form="cres" />
                        </measure>
                        <measure xml:id="measure-L5" n="2">
                            <staff xml:id="staff-L5F1" n="1">
                                <layer xml:id="layer-L5F1" n="1">
                                    <note xml:id="note-L6F1" dur="1" oct="4" pname="d" accid.ges="n" />
                                </layer>
                            </staff>
                        </measure>
                        <measure xml:id="measure-L7" n="3">
                            <staff xml:id="staff-L7F1" n="1">
                                <layer xml:id="layer-L7F1" n="1">
                                    <beam xml:id="beam-000000181289448">
                                        <note xml:id="note-L8F1" dur="8" oct="4" pname="e" accid.ges="n" />
                                        <note xml:id="note-L9F1" dur="8" oct="4" pname="f" accid.ges="n" />
                                    </beam>
                                    <space xml:id="space-L10F1" dots="1" dur="2" />
                                </layer>
                            </staff>
                            <hairpin xml:id="hairpin-000000192494598" staff="1" tstamp="1.000000" tstamp2="0m+1.5" form="cres" />
                        </measure>
                    </section>
                </score>
            </mdiv>
        </body>
    </music>
</mei>
```

Here is a sampler of hairpins with various angles:

<img width="698" alt="screen shot 2017-01-17 at 21 54 32" src="https://cloud.githubusercontent.com/assets/3487289/22052541/bb191c54-dcff-11e6-8c97-44d5379b019f.png">

I would say that the shortest two hairpins do not look good.  The shortest one looks best at around 19.9 degrees, and the second shortest looks best around 11.7 degress.  Not exactly the same value, but perhaps placing a limit of 20 degrees max for the hairpin angle might be good.

equation for calculating the angle (in degrees):

<img width="241" alt="screen shot 2017-01-17 at 22 01 41" src="https://cloud.githubusercontent.com/assets/3487289/22052706/94eaf088-dd00-11e6-8037-da9a8deba1e7.png">

Where *h* is the height and *w* is the width of the hairpin.

**Comment by craigsapp** [comment]
An exception is that adjacent cresc./decresc. need to have matching heights regardless of a maximum angle rule:


<img width="588" alt="screen shot 2017-01-17 at 22 25 29" src="https://cloud.githubusercontent.com/assets/3487289/22053183/e22b3922-dd03-11e6-9fc5-c7dcaac4fbe4.png">

**Comment by craigsapp** [comment]
Example in actual music:

<img width="1109" alt="screen shot 2017-01-17 at 22 26 53" src="https://cloud.githubusercontent.com/assets/3487289/22053216/18294c58-dd04-11e6-92db-a22650216751.png">

The crescendo hairpin in the third measure is shorter than the ones in the first two measures.

**Comment by lpugin** [comment]
Implemented in PR #702 and 8e72564e3ca9d244bfbc6eb2bea8ada0738c3434

![image](https://user-images.githubusercontent.com/689412/36601656-f316d180-18b5-11e8-82e6-4a616e6b457c.png)

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="http://music-encoding.org/schema/3.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="http://music-encoding.org/schema/3.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="3.0.0">
    <meiHead>
        <fileDesc>
            <titleStmt>
                <title />
            </titleStmt>
            <pubStmt>
                <date>2017-01-17 20:20:28</date>
            </pubStmt>
        </fileDesc>
        <encodingDesc>
            <projectDesc>
                <p>Transcoded from Humdrum with Verovio version 0.9.14-dev-fe01685</p>
            </projectDesc>
        </encodingDesc>
    </meiHead>
    <music>
        <body>
            <mdiv>
                <score>
                    <scoreDef xml:id="scoredef-000000201540999">
                        <staffGrp xml:id="m-000000150502018">
                            <staffDef xml:id="staffdef-000000183058674" clef.shape="G" clef.line="2" meter.count="4" meter.unit="4" n="1" lines="5" />
                        </staffGrp>
                    </scoreDef>
                    <section xml:id="section-000000015777043">
                        <measure xml:id="measure-L3" n="1">
                            <staff xml:id="staff-L3F1" n="1">
                                <layer xml:id="layer-L3F1" n="1">
                                    <note xml:id="note-L4F1" dur="1" oct="4" pname="c" accid.ges="n" />
                                </layer>
                            </staff>
                            <hairpin staff="1" tstamp="1.000000" tstamp2="1m+1.0" form="cres" />
                        </measure>
                        <measure xml:id="measure-L5" n="2">
                            <staff xml:id="staff-L5F1" n="1">
                                <layer xml:id="layer-L5F1" n="1">
                                    <note xml:id="note-L6F1" dur="1" oct="4" pname="d" accid.ges="n" />
                                </layer>
                            </staff>
                            <hairpin staff="1" tstamp="1.000000" tstamp2="0m+2.0" form="dim" />
                        </measure>
                        <measure xml:id="measure-L7" n="3">
                            <staff xml:id="staff-L7F1" n="1">
                                <layer xml:id="layer-L7F1" n="1">
                                    <beam xml:id="beam-000000181289448">
                                        <note xml:id="note-L8F1" dur="8" oct="4" pname="e" accid.ges="n" />
                                        <note xml:id="note-L9F1" dur="8" oct="4" pname="f" accid.ges="n" />
                                    </beam>
                                    <space xml:id="space-L10F1" dots="1" dur="2" />
                                </layer>
                            </staff>
                            <hairpin staff="1" tstamp="1.000000" tstamp2="0m+1.5" form="cres" />
                            <hairpin staff="1" tstamp="1.5" tstamp2="0m+1.8" form="dim" />
                        </measure>
                    </section>
                </score>
            </mdiv>
        </body>
    </music>
</mei>
```
